src/Microsoft.AspNet.Routing/AttributeRouting/TreeRouter.cs

```diff
@@ -0,0 +1,478 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Routing.Internal;
+using Microsoft.AspNet.Routing.Logging;
+using Microsoft.Extensions.Internal;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNet.Routing
+{
+    /// <summary>
+    /// An <see cref="IRouter"/> implementation for attribute routing.
+    /// </summary>
+    public class TreeRouter : IRouter
+    {
+        // Key used by routing and action selection to match an attribute route entry to a
+        // group of action descriptors.
+        public static readonly string RouteGroupKey = "!__route_group";
```

Comment:
Wasn't sure where to put this. I put it here because this is the only file it is used in Routing.